### ISSUE TITLE
ChannelRegistrar#unbindChannel doesn't free the channel.

### ISSUE THREAD

**Issue by BrainStone** [SpongePowered/SpongeForge#1818]
*(Previous title: ChannelRegistrar#getChannel returns empty Optional for registered channel)*

I registered a channel like this:
```java
Sponge.getChannelRegistrar().createRawChannel(plugin, "BungeeCord")
```
Then a later call to 
```java
Sponge.getChannelRegistrar().getChannel("BungeeCord")
```
returns an empty `Optional`.

This also causes `getOrCreateRaw` to throw the following exception when the channel is already registered:
```
org.spongepowered.api.network.ChannelRegistrationException: Error registering channel "BungeeCord" to Plugin{id=corelib, name=CoreLib, version=1.2.1.118-DEV, description=Core functions, a database and user API., authors=[BrainStone, WhoNeedsScript], source=C:\Users\Yannick\Desktop\Sponge-Testserver\mods\CoreLib-1.2.1.118-DEV.jar}
        at org.spongepowered.mod.network.SpongeModNetworkManager.createRawChannel(SpongeModNetworkManager.java:169) ~[SpongeModNetworkManager.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.api.network.ChannelRegistrar.getOrCreateRaw(ChannelRegistrar.java:122) ~[ChannelRegistrar.class:1.12.2-2503-7.0.0-BETA-2706]
        at de.project_creative.arrestia.corelib.util.network.Bungeecord.registerChannel(Bungeecord.java:67) ~[Bungeecord.class:1.2.1.118-DEV]
        at de.project_creative.arrestia.corelib.CoreLib.onGameInit(CoreLib.java:92) ~[CoreLib.class:1.2.1.118-DEV]
        at de.project_creative.arrestia.corelib.base.BasePlugin.lambda$onGameReload$6(BasePlugin.java:213) ~[BasePlugin.class:1.2.1.118-DEV]
        at de.project_creative.arrestia.corelib.base.BasePlugin.measureTime(BasePlugin.java:292) ~[BasePlugin.class:1.2.1.118-DEV]
        at de.project_creative.arrestia.corelib.base.BasePlugin.measureTime(BasePlugin.java:284) ~[BasePlugin.class:1.2.1.118-DEV]
        at de.project_creative.arrestia.corelib.base.BasePlugin.onGameReload(BasePlugin.java:198) ~[BasePlugin.class:1.2.1.118-DEV]
        at de.project_creative.arrestia.corelib.CoreLib.onGameReload(CoreLib.java:113) ~[CoreLib.class:1.2.1.118-DEV]
        at org.spongepowered.common.event.listener.GameReloadEventListener_CoreLib_onGameReload2.handle(Unknown Source) ~[?:?]
        at org.spongepowered.common.event.RegisteredListener.handle(RegisteredListener.java:95) ~[RegisteredListener.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.mod.event.SpongeModEventManager.post(SpongeModEventManager.java:316) ~[SpongeModEventManager.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.mod.event.SpongeModEventManager.post(SpongeModEventManager.java:349) ~[SpongeModEventManager.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.mod.event.SpongeModEventManager.post(SpongeModEventManager.java:333) ~[SpongeModEventManager.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.common.SpongeImpl.postEvent(SpongeImpl.java:198) ~[SpongeImpl.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.common.command.SpongeCommandFactory.lambda$createSpongePluginsCommand$7(SpongeCommandFactory.java:517) ~[SpongeCommandFactory.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.api.command.args.ChildCommandElementExecutor.execute(ChildCommandElementExecutor.java:181) [ChildCommandElementExecutor.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.api.command.spec.CommandSpec.process(CommandSpec.java:360) [CommandSpec.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.api.command.dispatcher.SimpleDispatcher.process(SimpleDispatcher.java:342) [SimpleDispatcher.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.common.command.SpongeCommandManager.process(SpongeCommandManager.java:321) [SpongeCommandManager.class:1.12.2-2503-7.0.0-BETA-2706]
        at net.minecraft.command.ServerCommandManager.func_71556_a(SourceFile:1083) [dh.class:?]
        at net.minecraft.server.dedicated.DedicatedServer.func_71333_ah(DedicatedServer.java:432) [nz.class:?]
        at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:397) [nz.class:?]
        at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:666) [MinecraftServer.class:?]
        at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:524) [MinecraftServer.class:?]
        at java.lang.Thread.run(Unknown Source) [?:1.8.0_144]
Caused by: java.lang.RuntimeException: That channel is already registered
        at net.minecraftforge.fml.common.network.NetworkRegistry.newChannel(NetworkRegistry.java:155) ~[NetworkRegistry.class:?]
        at org.spongepowered.mod.network.SpongeModChannelBinding.<init>(SpongeModChannelBinding.java:52) ~[SpongeModChannelBinding.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.mod.network.SpongeRawChannel.<init>(SpongeRawChannel.java:55) ~[SpongeRawChannel.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.mod.network.SpongeModNetworkManager.createRawChannel(SpongeModNetworkManager.java:167) ~[SpongeModNetworkManager.class:1.12.2-2503-7.0.0-BETA-2706]
        ... 25 more
```

**I am currently running...**
- SpongeForge version: 1.12.2-2503-7.0.0-BETA-2706
- Forge version: 14.23.0.2495
- Java version: 1.8.0_144
- Operating System:
 - [x] Linux (Ubuntu, Debian, Fedora, etc)
 - [ ] macOS
 - [x] Windows
- Plugins/Mods: Minecraft, Minecraft Coder Pack, SpongeAPI, SpongeForge, TestPlugin, LuckPerms, ViaBackwards, ViaVersion

**What is the expected result?**

`ChannelRegistrar#getChannel` should return the channel and `ChannelRegistrar#getOrCreateRaw` shouldn't throw the shown exception

**What is the current result?**

`ChannelRegistrar#getChannel`  returns an empty `Optional` and `ChannelRegistrar#getOrCreateRaw` throws the shown exception.

**Log**:  
https://gist.github.com/BrainStone/2877a3c43028cd2ef7638511ee4cb3c3

**Comment by simon816** [comment]
When creating the channel, are you sure no exceptions are thrown? Are you also sure that the channel is not being unbound (`unbindChannel`)?
`createRawChannel` followed immediately by `getChannel` will return the channel as expected.

**Comment by BrainStone** [comment]
Yes. And yes.  
I discovered that error while trying to unbind it.

It works fine under SpongeVanilla.

**Comment by simon816** [comment]
I see you have some other mods/plugins loaded. Are you able to isolate the problem to just your code that creates the channel? There could be another mod causing something strange to happen.

**Comment by BrainStone** [comment]
Even with only my plugin the same thing happens.

**Comment by simon816** [comment]
There may be an issue if you unbind a channel then attempt to create a new binding with the old name. Is that what could be happening here?
Otherwise, I don't know how it could be going wrong; creating a channel adds it to a `Map`, `getChannel` simply queries the map.

**Comment by BrainStone** [comment]
The getter fails before I unbind the channel.

I'll try to create a minimal example.

**Comment by BrainStone** [comment]
So this is a minimal example I could create:

```Java
import java.util.Collection;
import java.util.Optional;
import java.util.stream.Stream;
import org.spongepowered.api.Platform.Type;
import org.spongepowered.api.Platform;
import org.spongepowered.api.Sponge;
import org.spongepowered.api.event.Listener;
import org.spongepowered.api.event.game.GameReloadEvent;
import org.spongepowered.api.event.game.state.GameInitializationEvent;
import org.spongepowered.api.event.game.state.GameStoppingEvent;
import org.spongepowered.api.network.ChannelBinding;
import org.spongepowered.api.network.ChannelBuf;
import org.spongepowered.api.network.ChannelRegistrar;
import org.spongepowered.api.network.RawDataListener;
import org.spongepowered.api.network.RemoteConnection;
import org.spongepowered.api.plugin.Plugin;
import org.spongepowered.api.plugin.PluginContainer;
import com.google.inject.Inject;

@Plugin(id = "test", name = "Test", version = "0.0.0", description = "Test Plugin")
public class TestPlugin {
  @Inject
  private PluginContainer container;

  private static ChannelBinding.RawDataChannel channel;

  @Listener
  public void onGameInit(GameInitializationEvent event) {
    init();
  }

  @Listener
  public void onGameReload(GameReloadEvent event) {
    shutDown();

    init();
  }

  @Listener
  public void onGameStopping(GameStoppingEvent event) {
    shutDown();
  }

  private void init() {
    channel = Sponge.getChannelRegistrar().getOrCreateRaw(this, "BungeeCord");
  }

  private void shutDown() {
    final ChannelRegistrar channelRegistrar = Sponge.getChannelRegistrar();

    Stream.of(Type.CLIENT, Type.SERVER)
        .map(channelRegistrar::getRegisteredChannels)
        .flatMap(Collection::stream)
        .map(channelRegistrar::getChannel)
        .filter(Optional::isPresent)
        .map(Optional::get)
        .filter(channel -> container.getId().equals(channel.getOwner().getId()))
        .forEach(channelRegistrar::unbindChannel);
  }
}
```

Throws the exception when reloading:

```
[21:20:51] [Server thread/ERROR] [Sponge]: Could not pass GameReloadEvent$Impl to Plugin{id=test, name=Test, version=0.0.0, description=Test Plugin, source=C:\Sponge-Testserver\mods\Test-0.0.0.jar}
org.spongepowered.api.network.ChannelRegistrationException: Error registering channel "BungeeCord" to Plugin{id=test, name=Test, version=0.0.0, description=Test Plugin, source=C:\Sponge-Testserver\mods\Test-0.0.0.jar}
        at org.spongepowered.mod.network.SpongeModNetworkManager.createRawChannel(SpongeModNetworkManager.java:169) ~[SpongeModNetworkManager.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.api.network.ChannelRegistrar.getOrCreateRaw(ChannelRegistrar.java:122) ~[ChannelRegistrar.class:1.12.2-2503-7.0.0-BETA-2706]
        at TestPlugin.init(TestPlugin.java:47) ~[TestPlugin.class:?]
        at TestPlugin.onGameReload(TestPlugin.java:38) ~[TestPlugin.class:?]
        at org.spongepowered.common.event.listener.GameReloadEventListener_TestPlugin_onGameReload6.handle(Unknown Source) ~[?:?]
        at org.spongepowered.common.event.RegisteredListener.handle(RegisteredListener.java:95) ~[RegisteredListener.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.mod.event.SpongeModEventManager.post(SpongeModEventManager.java:316) ~[SpongeModEventManager.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.mod.event.SpongeModEventManager.post(SpongeModEventManager.java:349) ~[SpongeModEventManager.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.mod.event.SpongeModEventManager.post(SpongeModEventManager.java:333) ~[SpongeModEventManager.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.common.SpongeImpl.postEvent(SpongeImpl.java:198) ~[SpongeImpl.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.common.command.SpongeCommandFactory.lambda$createSpongePluginsCommand$7(SpongeCommandFactory.java:517) ~[SpongeCommandFactory.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.api.command.args.ChildCommandElementExecutor.execute(ChildCommandElementExecutor.java:181) [ChildCommandElementExecutor.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.api.command.spec.CommandSpec.process(CommandSpec.java:360) [CommandSpec.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.api.command.dispatcher.SimpleDispatcher.process(SimpleDispatcher.java:342) [SimpleDispatcher.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.common.command.SpongeCommandManager.process(SpongeCommandManager.java:321) [SpongeCommandManager.class:1.12.2-2503-7.0.0-BETA-2706]
        at net.minecraft.command.ServerCommandManager.func_71556_a(SourceFile:1083) [dh.class:?]
        at net.minecraft.server.dedicated.DedicatedServer.func_71333_ah(DedicatedServer.java:432) [nz.class:?]
        at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:397) [nz.class:?]
        at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:666) [MinecraftServer.class:?]
        at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:524) [MinecraftServer.class:?]
        at java.lang.Thread.run(Unknown Source) [?:1.8.0_144]
Caused by: java.lang.RuntimeException: That channel is already registered
        at net.minecraftforge.fml.common.network.NetworkRegistry.newChannel(NetworkRegistry.java:155) ~[NetworkRegistry.class:?]
        at org.spongepowered.mod.network.SpongeModChannelBinding.<init>(SpongeModChannelBinding.java:52) ~[SpongeModChannelBinding.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.mod.network.SpongeRawChannel.<init>(SpongeRawChannel.java:55) ~[SpongeRawChannel.class:1.12.2-2503-7.0.0-BETA-2706]
        at org.spongepowered.mod.network.SpongeModNetworkManager.createRawChannel(SpongeModNetworkManager.java:167) ~[SpongeModNetworkManager.class:1.12.2-2503-7.0.0-BETA-2706]
        ... 20 more
```

Works as expected on SpongeVanilla!

**Comment by BrainStone** [comment]
Some debugging revealed that the unbind call is sucessful. And the `ChannelRegistrar#getChannel` returns an empty optional. However the call to `ChannelRegistrar#createRawChannel` (`SpongeModNetworkManager#createRawChannel` in this case) hasn't realized that the channel has been freed again. So I suspect that the unbind method isn't working properly!

**Comment by simon816** [comment]
Ah, so as I suspected:
> There may be an issue if you unbind a channel then attempt to create a new binding with the old name.

SpongeForge doesn't actually remove the channel name from the Forge registry (I didn't want to hack around forge when I implemented it).
Didn't think anyone would rebind an old channel 😛 

I don't see why you would need to rebind a channel in reload anyway, unless it didn't bind the first time, there's no reason to remove it then re-add immediately.

But anyway, there is a bug with `unbindChannel`.
